zb-adapter.js

```diff
@@ -797,42 +815,56 @@ class ZigbeeAdapter extends Adapter {
 
   handleExplicitRx(frame) {
     if (this.zdo.isZdoFrame(frame)) {
-      this.zdo.parseZdoFrame(frame);
-      if (this.debugFrames) {
-        this.dumpFrame('Rcvd:', frame);
-      }
-      const clusterId = parseInt(frame.clusterId, 16);
-      if (clusterId in ZigbeeAdapter.zdoClusterHandler) {
-        ZigbeeAdapter.zdoClusterHandler[clusterId].call(this, frame);
-      } else {
-        console.error('No handler for ZDO cluster:',
-                      zdo.getClusterIdAsString(clusterId));
+      try {
+        this.zdo.parseZdoFrame(frame);
+        if (this.debugFrames) {
+          this.dumpFrame('Rcvd:', frame);
+        }
+        const clusterId = parseInt(frame.clusterId, 16);
+        if (clusterId in ZigbeeAdapter.zdoClusterHandler) {
+          ZigbeeAdapter.zdoClusterHandler[clusterId].call(this, frame);
+        } else {
+          console.error('No handler for ZDO cluster:',
+                        zdo.getClusterIdAsString(clusterId));
+        }
+      } catch (e) {
+        console.error('handleExplicitRx: Caught an exception parsing',
+                      'ZDO frame');
+        console.error(e);
+        console.error(frame);
       }
     } else if (this.isZhaFrame(frame)) {
-      zcl.parse(frame.data, parseInt(frame.clusterId, 16), (error, data) => {
-        if (error) {
-          console.error('Error parsing ZHA frame:', frame);
-          console.error(error);
-        } else {
-          frame.zcl = data;
-          if (this.debugFrames) {
-            this.dumpFrame('Rcvd:', frame);
-          }
-          // Add some special fields to ease waitFrame processing.
-          if (frame.zcl.seqNum) {
-            frame.zclSeqNum = frame.zcl.seqNum;
-          }
-          if (frame.zcl.cmdId) {
-            frame.zclCmdId = frame.zcl.cmdId;
-          }
-          const node = this.findNodeFromFrame(frame);
-          if (node) {
-            node.handleZhaResponse(frame);
+      try {
+        zcl.parse(frame.data, parseInt(frame.clusterId, 16), (error, data) => {
```

Comment:
Under what conditions can this throw? It looks like any errors get passed to the callback. Is this just for if the parseInt throws?

Reply:
I've only seen it throw once. I'm not sure why it threw because I didn't have packet tracing on, so I have no idea what the packet looked like.  The error I saw was that something was undefined, but I don't understand why. Adding this code exposes the packet that caused the error and then I can look and see why it threw.

Reply:
It threw from someplace inside the zcl.parse routine.

Reply:
Fascinating! Thanks for the context